sal/tests/server/unit/test_auth.py

```diff
@@ -0,0 +1,257 @@
+    is that a generated token can correctly verified 
+
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using the IP
+        AND a token generated less than the token lifetime ago
+        AND a token generated using the secret key
+        AND a token generated with a username
+    WHEN
+        The token is verified
+    THEN
+        The username is extracted from the token
+    """
+
+    # str cast required because hypothesis generates IPv4Address and
+    # IPv6Address objects
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':str(addr)}):
+        token = auth.generate_token(user)
+        assert user == auth.verify_token(token)
+
+
+def test_verify_expired_token(server_with_auth_prov):
+
+    """
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using the IP
+        AND a token generated more than the token lifetime ago
+        AND a token generated using the secret key
+        AND a token generated with a username
+    WHEN
+        The token is verified
+    THEN
+        None is returned
+    """
+
+    user = 'username'
+    server_with_auth_prov.config['SAL']['TOKEN_LIFETIME'] = 0
+
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':'0.0.0.0'}):
+        token = auth.generate_token(user)
+        sleep(1)
+        assert auth.verify_token(token) is None
+
+
+def test_verify_token_different_ip(server_with_auth_prov):
+
+    """
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using a different IP
+        AND a token generated more than the token lifetime ago
+        AND a token generated using the secret key
+        AND a token generated with a username
+    WHEN
+        The token is verified
+    THEN
+        None is returned
+    """
+
+    token_addr = '0.0.0.0'
+    request_addr = '20.313.236.111'
+    user = 'username'
+
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':token_addr}):
+        token = auth.generate_token(user)
+
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':str(request_addr)}):
+        assert auth.verify_token(token) is None
+
+
+def test_verify_invalid_token(server_with_auth_prov):
+
+    """
+    Tests that an invalid token (i.e., random, or generated with a different
+    secret key) is correctly verified
+
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using a different IP
+        AND a token generated more than the token lifetime ago
```

Comment:
**issue**: As with the above comment, this appears to be a copy/paste error
```suggestion
        AND a token generated using the IP
        AND a token generated less than the token lifetime ago
```

Reply:
Thanks!